File: 2023/6/solution.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.join(__dirname, 'input.txt'), 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
}).split('\n');

const numberPattern = /\d+/g;

const times = input[0].match(numberPattern).map(Number);
const distances = input[1].match(numberPattern).map(Number);

const numberOfWays = (time, distance) => {
  let j = 1;
  while (j < time && j * (time - j) <= distance) {
    j++;
  }
  let k = time - 1;
  while (k > 0 && k * (time - k) <= distance) {
    k--;
  }
  return k >= j ? k - j + 1 : 0;
}

let part1 = 1;

for (let i = 0; i < times.length; i++) {
  part1 *= numberOfWays(times[i], distances[i]);
}

console.log(part1);

const time2 = Number(input[0].match(numberPattern).reduce((a, b) => a + b));
const distance2 = Number(input[1].match(numberPattern).reduce((a, b) => a + b));

const part2 = numberOfWays(time2, distance2);
console.log(part2);

File: 2023/6/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	part1Input, part2Input := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(part1Input))
	fmt.Println(part2(part2Input))
}

func parseInput(data []string) ([][]int, []int) {
	numRegex := regexp.MustCompile(`\d+`)
	match0 := numRegex.FindAllString(data[0], -1)
	match1 := numRegex.FindAllString(data[1], -1)
	times := make([]int, len(match0))
	distances := make([]int, len(match1))
	for i, s := range match0 {
		n, err := strconv.Atoi(s)
		if err != nil {
			panic(err)
		}
		times[i] = n
	}
	for i, s := range match1 {
		n, err := strconv.Atoi(s)
		if err != nil {
			panic(err)
		}
		distances[i] = n
	}
	time2, err := strconv.Atoi(strings.Join(match0, ""))
	if err != nil {
		panic(err)
	}
	distance2, err := strconv.Atoi(strings.Join(match1, ""))
	if err != nil {
		panic(err)
	}
	return [][]int{times, distances}, []int{time2, distance2}
}

func waysToWin(time, distance int) int {
	var lo, hi int
	for (time-lo)*lo <= distance {
		lo++
	}
	hi = time
	for (time-hi)*hi <= distance {
		hi--
	}
	return hi - lo + 1
}

func part1(puzzleInput [][]int) int {
	res := 1
	for i := range puzzleInput[0] {
		res *= waysToWin(puzzleInput[0][i], puzzleInput[1][i])
	}
	return res
}

func part2(puzzleInput []int) int {
	return waysToWin(puzzleInput[0], puzzleInput[1])
}
